### ISSUE TITLE
pitest: increase mutation coverage for indentation profile to 100%

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#6294]
We should increase coverage for indentation profile up to 100%.
This issue is a subtask of #3708

**Comment by SGanguly1999** [comment]
I am on it.

**Comment by rohit9625** [comment]
Is somebody working on this issue?

**Comment by nrmancuso** [comment]
@rohit9625 no one is working on this issue, please be aware that this is most difficult pitest coverage issue we have. If you want to begin working on it, please kill one mutation per PR only.

**Comment by sktpy** [comment]
Is there any update that I might need to be aware of(asking since it has been a while since the issue was created), else I can start working on this issue.

**Comment by nrmancuso** [comment]
> Is there any thing that I might need to be aware of before I take this issue?(asking because it has been a while since this was created), else I can start working on it. 🚀

@sktpy 

1. ROI is low
2. Difficulty will be likely be high
3. Surviving mutations can be found at https://github.com/checkstyle/checkstyle/blob/72fa0536c643e56934cb7095061150ebf17a84e6/config/pitest-suppressions/pitest-indentation-suppressions.xml#L2

**Comment by sktpy** [comment]
> ROI is low

If that is the case, will focus on the issues that will be more impactful for our end users, but will visit this one periodically so that I can meet and greet boss level mutants. :space_invader:

**Comment by vivek-0509** [comment]
Hey @romani ,
While investigating PIT mutations in the indentation module, I encountered a suppressed mutation related to the call to primordialHandler.checkIndentation() in IndentationCheck.beginTree(). 
https://github.com/checkstyle/checkstyle/blob/aac12dfeb30a44cbc478826ecf1e400cb25216d1/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheck.java#L357-L359
On closer inspection, this call serves no functional purpose because PrimordialHandler.checkIndentation() has an empty implementation (// nothing to check). As a result, the mutation is equivalent: removing the call produces no observable behavior change, which explains why it cannot be killed by tests.
https://github.com/checkstyle/checkstyle/blob/aac12dfeb30a44cbc478826ecf1e400cb25216d1/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/PrimordialHandler.java#L36-L40

The cleanest solution is to remove this no operation call from beginTree(). However, doing so would reduce line coverage for PrimordialHandler.checkIndentation() because this method is not used anywhere and no integration test would reach that method anymore. Also the method itself cannot be removed, as it is required to satisfy the abstract contract defined by AbstractExpressionHandler.

To maintain coverage, I thought of adding a small test that directly invokes PrimordialHandler.checkIndentation().
This removes a suppressed PIT mutation, avoids dead code usage, and preserves full coverage.

Please let me know if adding a targeted unit test in this case is acceptable, given that Checkstyle generally prefers integration tests and avoids unit tests ?

**Comment by romani** [comment]
@vivek-0509 , yes, please send PR with pure unit test for this method, but make javadoc comment over such method to explain why we do unfortunate approach.

We have a pack of pure unit tests that clearly declares that they are 100% not functional, and serve to satisfy from static analysis tool.

**Comment by vivek-0509** [comment]
🎯 Mutation hunting progress update: The list is now half empty! Only 11 mutants left to kill. The hunt continues! 🚀

**Comment by romani** [comment]
thanks a lot for help.

you can use https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations#how-to-create-commit-that-skips-all-ci

and run diff report in PR, so if any diff happened, you will be able to copy such code to input to kill mutation

**Comment by vivek-0509** [comment]
Thank you for the advice. I will proceed with this approach.

**Comment by vivek-0509** [comment]
Hey @romani 
The three `VoidMethodCallMutator` mutations for `createdHandlers.clear()`, `handlerFactory.clearCreatedHandlers()`, and `handlers.clear()` are currently suppressed because these methods exist purely for memory management. They clear internal caches between file processing to prevent memory exhaustion when Checkstyle processes many files sequentially. For any single file, removing these calls produces  no difference, the check's logical behavior is unaffected.
These mutations survive because integration tests process files independently, and the only consequence of removing the `clear()` calls is an `OutOfMemoryError` after processing thousands of files, something impractical to test functionally.
However, we could potentially kill these mutations by adding **reflection-based unit tests** that directly verify the internal state is cleared. 
Would it be acceptable to add this type of reflection-based test to kill these mutations? Since this code does not cause any logical changes and exists solely for memory management, I wanted to confirm if such tests would be acceptable?

**Comment by romani** [comment]
We can add such method to be excluded from pitest mutators activity, see pom.xml in other profiles ,we already doing this for some methods.

**Comment by vivek-0509** [comment]
Hey @romani 
I investigated using excludedMethods as you suggested for the memory management mutations. Adding clearCreatedHandlers
to the exclusion list successfully prevents mutations inside the HandlerFactory.clearCreatedHandlers() method body.

However, the two mutations in IndentationCheck.beginTree() cannot be excluded this way. These are call site mutations where VoidMethodCallMutator removes the calls to handlerFactory.clearCreatedHandlers() and handlers.clear(). Since excludedMethods only protects the method body and not its callers, and both IndentationCheck and HandlerFactory are in the same target package, the call site mutations are still generated.
https://github.com/checkstyle/checkstyle/blob/8f16e4bdfb3851610b78f1a6dd72e98cd16017f8/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheck.java#L353-L358

To exclude these call site mutations, we would need to exclude beginTree entirely (which is not ideal since it contains important logic beyond memory cleanup).

Given these limitations, should I proceed with  reflection based tests to verify the clearing behavior or there is another approach that i should look for ?

**Comment by romani** [comment]
wrap:
```
handlerFactory.clearCreatedHandlers(); 
     handlers.clear(); 
```

with new method and exclude this new method.

**Comment by vivek-0509** [comment]
I actually tried wrapping the two clear calls into a new clearState() method and adding it to excludedMethods. While this successfully excludes the mutations inside the clearState() method body (the two clear calls),but a new mutation is still generated for the call to clearState() itself in beginTree(). The VoidMethodCallMutator can remove the entire clearState() call, which would still survive.

**Comment by vivek-0509** [comment]
@romani  ping

**Comment by romani** [comment]
Please send PR to let me see all details

**Comment by vivek-0509** [comment]
@romani I have raised the PR, there will be one surviving mutant please check.

**Comment by romani** [comment]
lets finish this issue in scope of https://github.com/checkstyle/checkstyle/issues/18035
